### PR TITLE
Dev basic architecture

### LINKED ISSUE #5: Set up base classes for scrapers/data collectors and Kafka producers

### PR DESCRIPTION
**Pull request by bastistrasser**

### CODE REVIEW

File: src/data_collection.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import logging
+from requests.exceptions import RequestException
+from constants import CONFIG_GENERAL, CONFIG_GENERAL_MAX_WORKERS, CONFIG_KAFKA, CONFIG_KAFKA_HOST, \
+    CONFIG_KAFKA_PORT, DATA_SOURCE_REDDIT, DATA_SOURCE_RSS, DATA_SOURCE_TWITTER
+from data_collectors import RedditDataCollector, RssDataCollector, TwitterDataCollector
+from producer import Producer
+
+
+def get_arguments():
+    parser = argparse.ArgumentParser(description='Script for collecting data from different data sources and publish it to a Kafka broker')
```
Comment (dnzxy) [inline src/data_collection.py]: *publishing

File: src/data_collectors.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import abc
+from concurrent.futures import Future
+from functools import partial
+import requests
+import re
+
+from constants import TIMEOUT_RSS_REQUEST
```
Comment (dnzxy) [inline src/data_collectors.py:5]: remove unnecessary line break
sort imports

File: src/data_collection.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+from functools import partial
+import logging
+from requests.exceptions import RequestException
+from constants import CONFIG_GENERAL, CONFIG_GENERAL_MAX_WORKERS, CONFIG_KAFKA, CONFIG_KAFKA_HOST, \
+    CONFIG_KAFKA_PORT, DATA_SOURCE_REDDIT, DATA_SOURCE_RSS, DATA_SOURCE_TWITTER
+from data_collectors import RedditDataCollector, RssDataCollector, TwitterDataCollector
+from producer import Producer
+
+
+def get_arguments():
```
Comment (dnzxy) [inline src/data_collection.py:13]: add doc string

File: src/data_collection.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    parser = argparse.ArgumentParser(description='Script for collecting data from different data sources and publish it to a Kafka broker')
+    parser.add_argument('--config', required=True, help='Configuration file for the data collection script')
+    subparsers = parser.add_subparsers(dest='data_source')
+    rss_parser = subparsers.add_parser('rss', help='Scrape data from RSS feeds')
+    rss_parser.add_argument('--base_url', required=True, help='URL of a RSS feed database where links to relevant RSS feeds can be found')
+    return parser.parse_args()
+
+
+def get_config(config_path):
```
Comment (dnzxy) [inline src/data_collection.py]: add doc string

File: src/data_collection.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    return parser.parse_args()
+
+
+def get_config(config_path): 
+    config = configparser.ConfigParser()
+    config.read(config_path)
+    return config
+
+
+def get_data_collector_instance(args):
```
Comment (dnzxy) [inline src/data_collection.py]: add doc string

File: src/data_collectors.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+from helper import get_request_with_timeout
+
+class BaseDataCollector(object):
```
Comment (dnzxy) [inline src/data_collectors.py]: add doc string

File: src/data_collectors.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+class BaseDataCollector(object): 
+    def __init__(self):
+        pass
+    
+    @abc.abstractmethod
+    def get_data_collection_futures(self, executor):
+        pass
+
+
+class RssDataCollector(BaseDataCollector):
```
Comment (dnzxy) [inline src/data_collectors.py]: add doc string
maybe with example what an instance of RssDataCollector and its properties would look like

Comment (dnzxy) [reply src/data_collectors.py]: also add doc strings to classes function and their respective params

File: src/helper.py [hunk 8]
```diff
@@ -0,0 +1,6 @@
+import requests
+
+def get_request_with_timeout(timeout):
```
Comment (dnzxy) [inline src/helper.py:4]: add doc string

File: src/producer.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
+from kafka import KafkaProducer
+import logging
+import uuid
+
+from constants import PRODUCER_API_VERSION
```
Comment (dnzxy) [inline src/producer.py:5]: remove unnecessary line break;
sort imports

File: src/data_collectors.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    
+
+    def get_data_collection_futures(self, executor):
+        feed_urls = self._get_feed_urls()
+        get_request = get_request_with_timeout(TIMEOUT_RSS_REQUEST)
+        futures = list(map(lambda url: executor.submit(get_request, url), feed_urls))
+        return futures
+
+    def _get_feed_urls(self):
+        headers = {
```
Comment (dnzxy) [inline src/data_collectors.py]: do these headers need to be editable / configurable? if so, maybe move them to some maintainable configuration or .env file?

File: src/producer.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+class Producer(object):
+    """Wrapper for a Kafka Producer
+    :param host: Hostname of the bootstrap server
```
Comment (dnzxy) [inline src/producer.py]: not entirely sure what methodology of a doc string this approach follows, but for readability's sake what do you think about changing this to something like:

```python

"""
Wrapper for a Kafka producer
    ----
    host (str): Hostname of the bootstrap server
    port (str): Port of bootstrap server
"""
```

Just an idea. I might very well just be unaware of this type of doc string annotation.

**Review by dnzxy — CHANGES_REQUESTED** [review]

File: src/producer.py (hunk 11 above)
Comment (bastistrasser) [reply src/producer.py]: https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html

File: src/data_collectors.py (hunk 2 above)
Comment (bastistrasser) [reply src/data_collectors.py:5]: https://peps.python.org/pep-0008/#imports

Imports should be grouped in the following order:

   1.  Standard library imports.
   2. Related third party imports.
   3. Local application/library specific imports.

**Review by lukasgraef — DISMISSED** [review]
Small problems with Issue #5 were fixed by bastis commit from 29/03/22

**Review by dnzxy — APPROVED** [review]

**Pull request merged**